File: leo2-ui/src/main/java/org/deku/leo2/fx/MainController.java
package org.deku.leo2.fx;

import javafx.animation.*;
import javafx.application.Platform;
import javafx.event.ActionEvent;
import javafx.event.EventHandler;
import javafx.fxml.FXML;
import javafx.fxml.Initializable;
import javafx.scene.Node;
import javafx.scene.control.Label;
import javafx.scene.control.ProgressIndicator;
import javafx.scene.layout.AnchorPane;
import javafx.scene.layout.Pane;
import javafx.util.Duration;
import org.deku.leo2.Main;
import org.deku.leo2.Settings;
import org.deku.leo2.bridge.LeoBridge;
import org.deku.leo2.bridge.Message;
import org.deku.leo2.fx.components.SidebarController;
import org.deku.leo2.fx.modules.DebugController;
import org.deku.leo2.fx.modules.DepotMaintenanceController;
import org.deku.leo2.fx.modules.HomeController;

import java.net.URL;
import java.util.ArrayList;
import java.util.ResourceBundle;

/**
 * Main userinterface controller
 * Created by masc on 21.09.14.
 */
public class MainController extends Controller implements Initializable, SidebarController.Listener, LeoBridge.Listener {
    @FXML
    private Label mTitle;
    @FXML
    private AnchorPane mContentPaneContainer;
    @FXML
    private Pane mSidebar;
    @FXML
    private SidebarController mSidebarController;
    @FXML
    private ProgressIndicator mProgressIndicator;

    private HomeController mHomeController;
    private DepotMaintenanceController mDepotMaintenanceController;
    private DebugController mDebugController;

    /**
     * Currently active module
     */
    private ModuleController mCurrentModuleController;
    /**
     * Progress indicator request count
     */
    private Integer mProgressIndicatorActivationCount = 0;

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        mSidebarController.setListener(this);
        LeoBridge.instance().getListenerEventDelegate().add(this);

        this.showModule(this.getHomeModule(), false);
    }

    /**
     * Tracks current content pane transition
     */
    Transition mContentPaneTransition;

    public DepotMaintenanceController getDepotMaintenanceModule() {
        if (mDepotMaintenanceController == null) {
            mDepotMaintenanceController = ModuleController.fromFxml("/fx/modules/DepotMaintenance.fxml");
        }
        return mDepotMaintenanceController;
    }

    public HomeController getHomeModule() {
        if (mHomeController == null) {
            mHomeController = ModuleController.fromFxml("/fx/modules/Home.fxml");
        }
        return mHomeController;
    }

    public DebugController getDebugPane() {
        if (mDebugController == null) {
            mDebugController = ModuleController.fromFxml("/fx/modules/Debug.fxml");
        }
        return mDebugController;
    }

    /**
     * Sets and shows a content pane
     *
     * @param moduleController
     * @param animated
     */
    private void showModule(ModuleController moduleController, boolean animated) {
        double duration = 500;

        if (mCurrentModuleController == moduleController)
            return;

        ModuleController oldModule = mCurrentModuleController;
        mCurrentModuleController = moduleController;

        Node node = moduleController.getRootNode();
        Node oldNode = (oldModule != null) ? oldModule.getRootNode() : null;

        this.setTitle(moduleController.getTitle(), animated);

        ArrayList<Animation> animations = new ArrayList<>();

        if (animated) {
            if (oldNode != null) {
                // Fade out old pane if there is one
                FadeTransition ftOut = new FadeTransition(Duration.millis(duration), oldNode);
                ftOut.setFromValue(1.0);
                ftOut.setToValue(0.0);
                ftOut.setOnFinished(e -> {
                    if (oldNode != mCurrentModuleController.getRootNode())
                        mContentPaneContainer.getChildren().remove(oldNode);
                });
                animations.add(ftOut);

                node.setOpacity(0.0);
            }
        } else {
            mContentPaneContainer.getChildren().clear();
        }

        // Add new pane to container
        if (!mContentPaneContainer.getChildren().contains(node))
            mContentPaneContainer.getChildren().add(node);
        mContentPaneContainer.setTopAnchor(node, 0.0);
        mContentPaneContainer.setBottomAnchor(node, 0.0);
        mContentPaneContainer.setRightAnchor(node, 0.0);
        mContentPaneContainer.setLeftAnchor(node, 0.0);

        if (animated) {
            // Fade in new pane
            FadeTransition ftIn = new FadeTransition(Duration.millis(duration), node);
            ftIn.setFromValue(0.0);
            ftIn.setToValue(1.0);
            ftIn.setOnFinished(e -> {
                mContentPaneTransition = null;
            });
            animations.add(ftIn);

            EventHandler<ActionEvent> evt = e -> {
                mContentPaneTransition = new ParallelTransition(animations.toArray(new Animation[0]));
                mContentPaneTransition.play();
            };

            if (mContentPaneTransition != null)
                mContentPaneTransition.setOnFinished(evt);
            else
                evt.handle(null);
        } else {
            node.setOpacity(1.0);
        }

        moduleController.activate();
    }

    public void showModule(ModuleController moduleController) {
        this.showModule(moduleController, Settings.instance().isAnimationsEnabled());
    }

    /**
     * Tracks current title transition
     */
    Transition mTitleTransition;

    /**
     * Set title
     *
     * @param title
     */
    private void setTitle(String title, boolean animated) {
        if (animated) {
            double duration = 175;

            FadeTransition ftOut = new FadeTransition(Duration.millis(duration), mTitle);
            ftOut.setFromValue(1.0);
            ftOut.setToValue(0.0);

            ftOut.setOnFinished(e -> {
                mTitle.setText(title);
            });

            FadeTransition ftIn = new FadeTransition(Duration.millis(duration), mTitle);
            ftIn.setFromValue(0.0);
            ftIn.setToValue(1.0);
            ftIn.setOnFinished(e -> {
                mTitleTransition = null;
            });

            // Create chained sequential transition
            EventHandler<ActionEvent> evt = e -> {
                mTitleTransition = new SequentialTransition(ftOut, ftIn);
                mTitleTransition.play();
            };

            if (mTitleTransition != null)
                mTitleTransition.setOnFinished(evt);
            else
                evt.handle(null);
        } else {
            mTitle.setText(title);
        }
    }

    /**
     * Request progress indication
     * Each call to request requires release to be called for the indicator to disappear as soon as all consumers released it
     */
    public void requestProgressIndicator() {
        mProgressIndicatorActivationCount++;
        mProgressIndicator.setVisible(true);
    }

    /**
     * Release progress indication
     */
    public void releaseProgressIndicator() {
        if (--mProgressIndicatorActivationCount <= 0) {
            mProgressIndicator.setVisible(false);
            mProgressIndicatorActivationCount = 0;
        }
    }

    @Override
    public void OnSidebarItemSelected(SidebarController.ItemType itemType) {
        ModuleController m;
        switch (itemType) {
            case Home:
                m = this.getHomeModule();
                break;
            case Depots:
                m = this.getDepotMaintenanceModule();
                break;
            case Debug:
                m = this.getDebugPane();
                break;
            default:
                throw new RuntimeException(String.format("Unknown sidebar item [%s]", itemType));
        }

        this.showModule(m);
    }

    @Override
    public void dispose() {
        super.dispose();

        if (mHomeController != null)
            mHomeController.dispose();
        if (mDepotMaintenanceController != null)
            mDepotMaintenanceController.dispose();
        if (mDebugController != null)
            mDebugController.dispose();
    }

    @Override
    public void onLeoBridgeMessageReceived(Message message) {
        Platform.runLater(() ->
        {
            Main.instance().toForeground();
            if (message.get("view").equals("depot")) {
                Integer id = (Integer) message.get("id");
                this.showModule(this.getDepotMaintenanceModule());
                this.getDepotMaintenanceModule().selectDepot(id);
            } else {
                Main.instance().showMessage(String.format("Received message [%s]", message));
            }
        });
    }
}

File: leoz-ui/src/main/java/org/deku/leoz/fx/MainController.kt
package org.deku.leoz.fx

import com.github.salomonbrys.kodein.Kodein
import com.github.salomonbrys.kodein.conf.global
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.lazy
import javafx.animation.*
import javafx.application.Platform
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.Initializable
import javafx.scene.control.Label
import javafx.scene.control.ProgressIndicator
import javafx.scene.layout.AnchorPane
import javafx.scene.layout.Pane
import javafx.util.Duration
import org.deku.leoz.Main
import org.deku.leoz.Settings
import org.deku.leoz.bridge.LeoBridge
import org.deku.leoz.bridge.Message
import org.deku.leoz.fx.components.SidebarController
import org.deku.leoz.fx.modules.DebugController
import org.deku.leoz.fx.modules.DepotMaintenanceController
import org.deku.leoz.fx.modules.HomeController
import java.net.URL
import java.util.*

/**
 * Main userinterface controller
 * Created by masc on 21.09.14.
 */
class MainController : Controller(), Initializable, SidebarController.Listener, LeoBridge.Listener {
    @FXML
    private lateinit var fxTitle: Label
    @FXML
    private lateinit var fxVersion: Label
    @FXML
    private lateinit var fxContentPaneContainer: AnchorPane
    @FXML
    private lateinit var fxSidebar: Pane
    @FXML
    private lateinit var fxSidebarController: SidebarController
    @FXML
    private lateinit var fxProgressIndicator: ProgressIndicator

    private val settings: Settings by Kodein.global.lazy.instance()

    val homeController by lazy {
        ModuleController.fromFxml<HomeController>("/fx/modules/Home.fxml")
    }

    val depotMaintenanceController by lazy {
        ModuleController.fromFxml<DepotMaintenanceController>("/fx/modules/DepotMaintenance.fxml")
    }

    val debugController by lazy {
        ModuleController.fromFxml<DebugController>("/fx/modules/Debug.fxml")
    }

    /**
     * Currently active module
     */
    private var currentModuleController: ModuleController? = null
    /**
     * Progress indicator request count
     */
    private var progressIndicatorActivationCount: Int = 0

    /**
     * Tracks current content pane transition
     */
    private var contentPaneTransition: Transition? = null

    override fun initialize(location: URL, resources: ResourceBundle) {
        fxSidebarController.setListener(this)
        LeoBridge.instance().listenerEventDelegate.add(this)

        this.showModule(this.homeController, false)
        this.setVersion("User: Max Mustermann\tVersion: TEST") //TODO: Dynamic information
    }

    /**
     * Sets and shows a content pane

     * @param moduleController
     * *
     * @param animated
     */
    private fun showModule(moduleController: ModuleController, animated: Boolean) {
        val duration = 500.0

        if (currentModuleController === moduleController)
            return

        val oldModule = currentModuleController
        currentModuleController = moduleController

        val node = moduleController.rootNode
        val oldNode = oldModule?.rootNode

        this.setTitle(moduleController.title, animated)

        val animations = ArrayList<Animation>()

        if (animated) {
            if (oldNode != null) {
                // Fade out old pane if there is one
                val ftOut = FadeTransition(Duration.millis(duration), oldNode)
                ftOut.fromValue = 1.0
                ftOut.toValue = 0.0
                ftOut.setOnFinished { e ->
                    if (oldNode !== currentModuleController!!.rootNode)
                        fxContentPaneContainer.children.remove(oldNode)
                }
                animations.add(ftOut)

                node!!.opacity = 0.0
            }
        } else {
            fxContentPaneContainer.children.clear()
        }

        // Add new pane to container
        if (!fxContentPaneContainer.children.contains(node))
            fxContentPaneContainer.children.add(node)
        AnchorPane.setTopAnchor(node, 0.0)
        AnchorPane.setBottomAnchor(node, 0.0)
        AnchorPane.setRightAnchor(node, 0.0)
        AnchorPane.setLeftAnchor(node, 0.0)

        if (animated) {
            // Fade in new pane
            val ftIn = FadeTransition(Duration.millis(duration), node)
            ftIn.fromValue = 0.0
            ftIn.toValue = 1.0
            ftIn.setOnFinished { e -> contentPaneTransition = null }
            animations.add(ftIn)

            val evt: EventHandler<ActionEvent> = EventHandler {
                contentPaneTransition = ParallelTransition(*animations.toTypedArray())
                contentPaneTransition!!.play()
            }

            if (contentPaneTransition != null)
                contentPaneTransition!!.onFinished = evt
            else
                evt.handle(null)
        } else {
            node!!.opacity = 1.0
        }

        moduleController.activate()
        fxSidebarController.highlightByController(moduleController)
    }

    fun showModule(moduleController: ModuleController) {
        this.showModule(moduleController, this.settings.isAnimationsEnabled)
    }

    /**
     * Tracks current title transition
     */
    internal var titleTransition: Transition? = null

    /**
     * Set title

     * @param title
     */
    private fun setTitle(title: String, animated: Boolean) {
        if (animated) {
            val duration = 175.0

            val ftOut = FadeTransition(Duration.millis(duration), fxTitle)
            ftOut.fromValue = 1.0
            ftOut.toValue = 0.0

            ftOut.setOnFinished { e -> fxTitle.text = title }

            val ftIn = FadeTransition(Duration.millis(duration), fxTitle)
            ftIn.fromValue = 0.0
            ftIn.toValue = 1.0
            ftIn.setOnFinished { e -> titleTransition = null }

            // Create chained sequential transition
            val evt: EventHandler<ActionEvent> = EventHandler {
                titleTransition = SequentialTransition(ftOut, ftIn)
                titleTransition!!.play()
            }

            if (titleTransition != null)
                titleTransition!!.onFinished = evt
            else
                evt.handle(null)
        } else {
            fxTitle.text = title
        }
    }

    private fun setVersion(version: String){
        fxVersion.text = version
    }

    /**
     * Request progress indication
     * Each call to request requires release to be called for the indicator to disappear as soon as all consumers released it
     */
    fun requestProgressIndicator() {
        progressIndicatorActivationCount += 1
        fxProgressIndicator.isVisible = true
    }

    /**
     * Release progress indication
     */
    fun releaseProgressIndicator() {
        progressIndicatorActivationCount -= 1
        if (progressIndicatorActivationCount <= 0) {
            fxProgressIndicator.isVisible = false
            progressIndicatorActivationCount = 0
        }
    }

    override fun OnSidebarItemSelected(itemType: SidebarController.ItemType) {
        val m: ModuleController
        when (itemType) {
            SidebarController.ItemType.Home -> m = this.homeController
            SidebarController.ItemType.Depots -> m = this.depotMaintenanceController
            SidebarController.ItemType.Debug -> m = this.debugController
            else -> throw RuntimeException(String.format("Unknown sidebar item [%s]", itemType))
        }

        this.showModule(m)
    }

    override fun close() {
        super.close()

        if (homeController != null)
            homeController!!.close()
        if (depotMaintenanceController != null)
            depotMaintenanceController!!.close()
        if (debugController != null)
            debugController!!.close()
    }

    override fun onLeoBridgeMessageReceived(message: Message) {
        Platform.runLater {
            if (message.get("view") == "depot") {
                val id = message.get("id") as Int
                this.showModule(this.depotMaintenanceController)
                this.depotMaintenanceController.selectDepot(id)
            } else {
                Main.instance().showMessage(String.format("Received message [%s]", message))
            }
            Main.instance().toForeground()
        }
    }
}
